mitodl/mitxonline
# Adds updated dashboard UI for programs

#### Pre-Flight checklist


#### What are the relevant tickets?

Fixes #1040 

#### What's this PR do?

This PR:
- Refactors a good bit of the dashboard code to clean it up and isolate a few repeated things in the code
- With the program UI enabled (`FEATURE_ENABLE_PROGRAM_UI`):
   - Adds a tab switcher to the top of the dashboard page
   - Adds a My Courses tab that replicates the existing dashboard functionality - it displays all of the courses that the learner has enrolled in
   - Adds a Programs tab that displays _just_ the programs that the learner is enrolled in
   - For each rendered program, allows the learner to click to expand the program drawer, which will display all the courses in the program, grouped by the learner's enrollment status (either enrolled or not) 

#### How should this be manually tested?

Make sure you have the feature flag turned on (or append `?enabled_programs` to your URL) and that you have a learner account with a smattering of enrollments. (For best results, the learner should be enrolled in courses in multiple programs, with programs that have additional courses the learner isn't enrolled in, and stand-alone courses.) Then, view the dashboard. You should be able to switch between tabs, open the drawer on the Program tab, and the enrolled courses should work as they did before (including the menu and unenroll/email settings options). Program courses without an enrollment should appear in the Available section in the drawer and should allow you to navigate to the course detail page, but should not have options for enrollment. 

#### Screenshots (if appropriate)

My Courses tab:
![image](https://user-images.githubusercontent.com/945611/193913660-b0805926-3cf5-4347-8ab8-3be2fb8074d7.png)
Programs tab:
![image](https://user-images.githubusercontent.com/945611/193913731-ba1725ce-20e9-4eb1-a64d-7313a4f295f0.png)
Program drawer:
![image](https://user-images.githubusercontent.com/945611/193913703-fb0af999-1d51-446a-84bd-3ad126c97b45.png)

My Courses with an upgradable enrollment:
![image](https://user-images.githubusercontent.com/945611/193914386-ce2af325-695c-4dc4-970e-d0b41e69b690.png)
Progam drawer with an upgradable enrollment:
![image](https://user-images.githubusercontent.com/945611/193914440-fb3528b6-7a32-4d8b-b4a5-1cbee8ddeac1.png)


#### Other

- Not sure on the badging for the courses - I thought we'd eschewed that? Can add it back in pretty easily.
- Per the original issue, this does not sort the program courses by status beyond enrolled/not-enrolled, nor does it include the program focus (as that's not a field we have).
